src/Compilers/Core/Portable/MetadataReader/PEAssembly.cs

```diff
@@ -63,19 +61,28 @@ internal PEAssembly(AssemblyMetadata owner, ImmutableArray<PEModule> modules)
 
             _identity = modules[0].ReadAssemblyIdentityOrThrow();
 
-            var refs = ArrayBuilder<AssemblyIdentity>.GetInstance();
-            int[] refCounts = new int[modules.Length];
+            // Pre-calculate size to allow this code to only require a single ImmutableArray allocation.
+            var totalRefCount = 0;
+            var refCounts = ImmutableArray.CreateBuilder<int>(modules.Length);
+
+            for (int i = 0; i < modules.Length; i++)
+            {
+                ImmutableArray<AssemblyIdentity> refsForModule = modules[i].ReferencedAssemblies;
+                refCounts.Add(refsForModule.Length);
+                totalRefCount += refsForModule.Length;
+            }
```

Comment:
💡 Use `Sum` to calculate `totalRefCount`, and move the `refCounts.Add` line down into the other loop where `refCounts[i]` was before